File: C++/Medium/Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_set<char> seen;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[i].size(); j++) {
                if (board[i][j] != '.' && seen.contains(board[i][j])) {
                    return false;
                }
                seen.insert(board[i][j]);
            }
            seen.clear();
        }

        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board.size(); j++) {
                if (board[j][i] != '.' && seen.contains(board[j][i])) {
                    return false;
                }
                seen.insert(board[j][i]);
            }
            seen.clear();
        }

        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                for (int k = 0; k < 3; k++) {
                    for (int l = 0; l < 3; l++) {
                        if (board[(i * 3) + k][(j * 3) + l] != '.' && seen.contains(board[(i * 3) + k][(j * 3) + l])) {
                            return false;
                        }
                        seen.insert(board[(i * 3) + k][(j * 3) + l]);
                    }
                }
                seen.clear();
            }
        }


        return true;

    }
};

File: Python/Medium/Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        BOARD_SIZE = 9
        BOX_SIZE = 3;
        seen = set()
        for row in range(9):
            for col in range(9):
                char = board[row][col]
                if char in seen and char != ".":
                    return False
                seen.add(char)
            seen.clear()

        for col in range(9):
            for row in range(9):
                char = board[row][col]
                if char in seen and char != ".":
                    return False
                seen.add(char)
            seen.clear()

        for box_row in range(0, 9, 3):
            for box_col in range(0, 9, 3):
                for row in range(box_row, box_row + BOX_SIZE):
                    for col in range(box_col, box_col + BOX_SIZE):
                        char = board[row][col]
                        if char in seen and char != ".":
                            return False
                        seen.add(char)
                seen.clear()
        return True


    # there are methods that use less time complexity but more space (example one pass but 27 sets(9 for rows, 9 for cols, 9 for boxes))